content/Link Tracking with StimulusReflex - Part I.md

```diff
@@ -0,0 +1,152 @@
+
+First, we'll create a basic controller with an index action to just render some HTML
+
+```bash
+rails g controller pages index
+```
+
+And we'll make this action our new root
+
+```ruby
+# change
+get 'pages/index'
+
+# to
+root 'pages#index'
+```
+
+And we'll dive into some Javascript as well by changing the `HelloController` in `app/javascript/controllers/hello_controller.js` for the following class
+
+```javascript
+import { Controller } from "stimulus"
+
+export default class extends Controller {
+  static targets = ["output"]
+
+  sayHello() {
+    this.outputTarget.textContent = "Hello, Stimulus!"
+  }
+}
+```
+
+Now we just need to bind it to our markup in `app/views/pages/index.html.erb` by using some
+data attributes
+
+```html
+<div data-controller="hello">
+  <button data-action="click->hello#sayHello">Say hello</button>
+  <h1 data-target="hello.output"></h1>
+</div>
+```
+
+Now if you run the server with `rails s` and go to `http://localhost:3000` you'll see that when you click on the "Say hello" button you'll get back "Hello, Stimulus!". I think the code it's pretty self explanatory but I'll give a quick walkthrough anyway.
+
+Here we are just telling StimulusJS to hook the controller prefixed with `Hello` scoped to that HTML tree. That will start looking for specific data attributes which in this case are `data-action` and `data-target`. The first accepts a syntax to call specific methods on the instance of the controller on a given event, currently set to invoke the `sayHello` method when the click event occurs. The latter binds the text to a target, which is just like a reference to the element but more at the same time, keeping them synchronized.
+
+So in this case when we set the `textContent` of our `outputTarget` the text will immediately appear on our `h1` tag. Keep this concept in the back of your mind since we'll keep using it throughout the series.
+
+## Adding Reflex to our previous example
+
+Now that we've seen how StimulusJS works let's get down to business. The first step it's to install it, which is extremely simple
+
+```bash
+bundle add stimulus_reflex
+bundle exec rails stimulus_reflex:install
+```
+
+Then let's do some changes to the generated code by renaming `app/reflexes/example_reflex.rb` to `app/reflexes/hello_reflex.rb` and adjust the code inside as well
+
+
+```ruby
+# change
+class ExampleReflex < ApplicationReflex
+
+# to
+class HelloReflex < ApplicationReflex
+```
+
+And add a handler method called `#greet` to differentiate it from our frontend binding
+
+```ruby
+  def greet
+    @message = 'Hello, StimulusReflex!'
+  end
+```
+
+And change the `h1` tag in `app/views/pages/index.html.erb` to show our message
+
+```html
+<h1><%= @message %></h1>
+```
+
+You *DO NOT* need to add the `@message` variable in the `PagesController`. If you do set it remember to use the `||=` operator like `@message ||= "Some default"` since otherwise it would override the Reflex's value.
```

Comment:
I would remove the `do` here from `If you do set...`

Reply:
How do you mean?